crates/ide-assists/src/handlers/convert_let_else_to_match.rs

```diff
@@ -0,0 +1,426 @@
+        OrPat(p) => {
+            let mut v = vec![];
+            for p in p.pats() {
+                v.extend(binders_in_pat(&p)?);
+            }
+            Some(v)
+        }
+        ParenPat(p) => p.pat().and_then(|p| binders_in_pat(&p)),
+        RangePat(p) => {
+            let mut start = if let Some(st) = p.start() { binders_in_pat(&st)? } else { vec![] };
+            let end = if let Some(st) = p.end() { binders_in_pat(&st)? } else { vec![] };
+            start.extend(end);
+            Some(start)
+        }
+        RecordPat(p) => {
+            let mut v = vec![];
+            for f in p.record_pat_field_list()?.fields() {
+                let pat = f.pat()?;
+                v.extend(binders_in_pat(&pat)?);
+            }
+            Some(v)
+        }
+        RefPat(p) => p.pat().and_then(|p| binders_in_pat(&p)),
+        SlicePat(p) => {
+            let mut v = vec![];
+            for p in p.pats() {
+                v.extend(binders_in_pat(&p)?);
+            }
+            Some(v)
+        }
+        TuplePat(p) => {
+            let mut v = vec![];
+            for p in p.fields() {
+                v.extend(binders_in_pat(&p)?);
+            }
+            Some(v)
+        }
+        TupleStructPat(p) => {
+            let mut v = vec![];
+            for p in p.fields() {
+                v.extend(binders_in_pat(&p)?);
+            }
+            Some(v)
+        }
+        // don't support macro pat yet
+        MacroPat(_) => None,
+    }
+}
+
+fn binders_to_str(binders: &[(String, bool)], addmut: bool) -> String {
+    let vars = binders
+        .iter()
+        .map(
+            |(ident, ismut)| {
+                if *ismut && addmut {
+                    format!("mut {}", ident)
+                } else {
+                    ident.to_string()
+                }
+            },
+        )
+        .collect::<Vec<_>>()
+        .join(", ");
+    if binders.is_empty() {
+        String::from("{}")
+    } else if binders.len() == 1 {
+        vars
+    } else {
+        format!("({})", vars)
+    }
+}
+
+// Assist: convert_let_else_to_match
+//
+// Converts let-else statement to let statement and match expression.
+//
+// ```
+// fn main() {
+//     let Ok(mut x) = f() else {$0 return };
+// }
+// ```
+// ->
+// ```
+// fn main() {
+//     let mut x = match f() {
+//         Ok(x) => x,
+//         _ => return,
+//     };
+// }
+// ```
+pub(crate) fn convert_let_else_to_match(acc: &mut Assists, ctx: &AssistContext) -> Option<()> {
+    let let_stmt: LetStmt = ctx.find_node_at_offset()?;
+    let let_else_block = let_stmt.let_else()?.block_expr()?;
+    let let_init = let_stmt.initializer()?;
+    if let_stmt.ty().is_some() {
+        // don't support let with type annotation
+        return None;
+    }
+    let pat = let_stmt.pat()?;
+    let binders = binders_in_pat(&pat)?;
+
+    let target = let_stmt.syntax().text_range();
```

Comment:
We generally try to keep assists scoped a bit more narrowly than this, because this makes it show up even when pretty deep inside other code in the let-else block. I think it would be best if this triggered when the cursor is on the `else` token only.

Reply:
That's a good point. I've updated the PR.